dali/operators/image/remap/cvcuda/warp_perspective.cc

```diff
@@ -0,0 +1,241 @@
+      return fill_value;
+    } else {
+      return float4{0, 0, 0, 0};
+    }
+  }
+
+  void ValidateTypes(const Workspace &ws) const {
+    auto inp_type = ws.Input<GPUBackend>(0).type();
+    DALI_ENFORCE(inp_type == DALI_UINT8 || inp_type == DALI_INT16 || inp_type == DALI_UINT16 ||
+                 inp_type == DALI_FLOAT,
+                 "The operator accepts the following input types: "
+                 "uint8, int16, uint16, float.");
+    if (ws.NumInput() > 1) {
+      auto mat_type = ws.Input<GPUBackend>(1).type();
+      DALI_ENFORCE(mat_type == DALI_FLOAT,
+                   "Transformation matrix can be provided only as float32 values.");
+    }
+  }
+
+  bool OCVCompatArg(const std::string &arg) {
+    if (arg == "corner") {
+      return false;
+    } else if (arg == "center") {
+      return true;
+    } else {
+      DALI_FAIL(make_string("Invalid pixel_origin argument: ", arg));
+    }
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) override {
+    ValidateTypes(ws);
+    const auto &input = ws.Input<GPUBackend>(0);
+    auto input_shape = input.shape();
+    auto input_layout = input.GetLayout();
+    output_desc.resize(1);
+
+    auto output_shape = input_shape;
+    int channels = (input_layout.find('C') != -1) ? input_shape[0][input_layout.find('C')] : -1;
+    fill_value_ = GetFillValue(channels);
+    if (size_arg_.HasExplicitValue()) {
+      size_arg_.Acquire(spec_, ws, input_shape.size(), TensorShape<1>(2));
+      for (int i = 0; i < input_shape.size(); i++) {
+        auto height = std::max<int>(std::roundf(size_arg_[i].data[0]), 1);
+        auto width = std::max<int>(std::roundf(size_arg_[i].data[1]), 1);
+        auto out_sample_shape = (channels != -1) ? TensorShape<>({height, width, channels}) :
+                                                   TensorShape<>({height, width});
+        output_shape.set_tensor_shape(i, out_sample_shape);
+      }
+    }
+
+    output_desc[0] = {output_shape, input.type()};
+    return true;
+  }
+
+  void RunImpl(Workspace &ws) override {
+    const auto &input = ws.Input<GPUBackend>(0);
+    auto &output = ws.Output<GPUBackend>(0);
+    output.SetLayout(input.GetLayout());
+
+    kernels::DynamicScratchpad scratchpad({}, AccessOrder(ws.stream()));
+
+    nvcv::Tensor matrix{};
+    if (ws.NumInput() > 1) {
+      DALI_ENFORCE(!matrix_arg_.HasExplicitValue(),
+                   "Matrix input and `matrix` argument should not be provided at the same time.");
+      auto &matrix_input = ws.Input<GPUBackend>(1);
+      DALI_ENFORCE(is_uniform(matrix_input.shape()),
+                   "Matrix data has to be a uniformly shaped batch.");
+      DALI_ENFORCE(matrix_input.tensor_shape(0) == TensorShape<2>(3, 3),
+                   make_string("Transformation matrix must be a 3x3 matrix. "
+                               "Instead got data with shape: {",
+                               matrix_input.tensor_shape(0), "}"));
```

Comment:
Not very user friendly - first we raise one error and then possibly another, for essentially the same thing.
How about:
```suggestion
      DALI_ENFORCE(is_uniform(matrix_input.shape()),
                   "Matrix data has to be a uniformly shaped batch.");
      DALI_ENFORCE(matrix_input.shape() == uniform_list_shape(matrix_input.num_samples(), TensorShape<2>(3, 3)),
                   make_string("Expected a uniform list of 3x3 matrices. "
                               "Instead got data with shape: ",
                               matrix_input.tensor_shape()));
```

Reply:
Done